internal/rm/health.go

```diff
@@ -17,189 +17,460 @@
+	// specified here override the ones specified in the
+	// `DP_DISABLE_HEALTHCHECKS`. Note that this also allows individual
+	// XIDs to be selected when ALL XIDs are disabled.
 	envEnableHealthChecks = "DP_ENABLE_HEALTHCHECKS"
 )
 
-// CheckHealth performs health checks on a set of devices, writing to the 'unhealthy' channel with any unhealthy devices
-func (r *nvmlResourceManager) checkHealth(stop <-chan interface{}, devices Devices, unhealthy chan<- *Device) error {
-	xids := getDisabledHealthCheckXids()
-	if xids.IsAllDisabled() {
+// HealthProvider manages GPU device health monitoring with lifecycle
+// control.
+type HealthProvider interface {
+	// Start initiates health monitoring. Blocks until initial setup
+	// completes. Returns error if health monitoring cannot be started.
+	Start(context.Context) error
+
+	// Stop gracefully shuts down health monitoring and waits for all
+	// goroutines to complete. Safe to call multiple times.
+	Stop()
+
+	// Health returns a read-only channel that receives devices that
+	// have become unhealthy.
+	Health() <-chan *Device
+}
+
+// nvmlHealthProvider implements HealthProvider using NVML event
+// monitoring. This is a refactoring of the existing checkHealth logic
+// with proper lifecycle management.
+type nvmlHealthProvider struct {
+	// Configuration
+	nvml    nvml.Interface
+	config  *spec.Config
+	devices Devices
+
+	// NVML resources
+	eventSet nvml.EventSet
+
+	// Lifecycle management
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+
+	// State guards
+	mu      sync.Mutex
+	started bool
+	stopped bool
+
+	// Communication
+	healthChan chan *Device
+
+	// XID filtering
+	xidsDisabled disabledXIDs
+
+	// Device placement maps (for MIG support)
+	parentToDeviceMap map[string]*Device
+	deviceIDToGiMap   map[string]uint32
+	deviceIDToCiMap   map[string]uint32
+}
+
+// NewNVMLHealthProvider creates a new health provider for NVML devices.
+// Does not start monitoring - caller must call Start().
+func NewNVMLHealthProvider(
+	nvml nvml.Interface,
+	config *spec.Config,
+	devices Devices,
+) HealthProvider {
+	return &nvmlHealthProvider{
+		nvml:       nvml,
+		config:     config,
+		devices:    devices,
+		healthChan: make(chan *Device, 64),
+	}
+}
+
+// Start initializes NVML, registers event handlers, and starts the
+// monitoring goroutine. Blocks until initialization completes.
+func (p *nvmlHealthProvider) Start(ctx context.Context) error {
+	p.mu.Lock()
+	if p.started {
+		p.mu.Unlock()
+		return fmt.Errorf("health provider already started")
+	}
+	p.started = true
+	p.mu.Unlock()
+
+	// Check if health checks are disabled
+	p.xidsDisabled = getDisabledHealthCheckXids()
+	if p.xidsDisabled.IsAllDisabled() {
+		klog.Info("Health checks disabled via DP_DISABLE_HEALTHCHECKS")
 		return nil
 	}
 
-	ret := r.nvml.Init()
+	// Initialize NVML
+	ret := p.nvml.Init()
 	if ret != nvml.SUCCESS {
-		if *r.config.Flags.FailOnInitError {
+		if *p.config.Flags.FailOnInitError {
```

Comment:
nit: Let's not rename `r` to `p` in a single commit. (see comment on managing diffs).

Reply:
thanks for the suggestion, adopted , now an independent commit